cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/ZookeeperReconcilerKRaftMigrationTest.java

```diff
@@ -0,0 +1,216 @@
+
+        Kafka patchedKafka = new KafkaBuilder(KAFKA)
+                .editMetadata()
+                    .addToAnnotations(Annotations.ANNO_STRIMZI_IO_KRAFT, "enabled")
+                .endMetadata()
+                .withNewStatus()
+                    .withKafkaMetadataState(KafkaMetadataState.KRaft.name())
+                .endStatus()
+                .build();
+
+        ResourceOperatorSupplier supplier = ResourceUtils.supplierWithMocks(false);
+
+        SecretOperator mockSecretOps = supplier.secretOperations;
+        Secret secret = new Secret();
+        when(mockSecretOps.getAsync(eq(NAMESPACE), eq(KafkaResources.clusterCaCertificateSecretName(CLUSTER_NAME)))).thenReturn(Future.succeededFuture(secret));
+        when(mockSecretOps.getAsync(eq(NAMESPACE), eq(KafkaResources.secretName(CLUSTER_NAME)))).thenReturn(Future.succeededFuture(secret));
+        KafkaVersionChange versionChange = new KafkaVersionChange(VERSIONS.defaultVersion(), VERSIONS.defaultVersion(), VERSIONS.defaultVersion().protocolVersion(), VERSIONS.defaultVersion().messageVersion(), VERSIONS.defaultVersion().metadataVersion());
+
+        KafkaMetadataStateManager stateManager = new KafkaMetadataStateManager(RECONCILIATION, patchedKafka, CO_CONFIG.featureGates().useKRaftEnabled());
+
+        MockZooKeeperReconciler zookeeperReconciler = spy(new MockZooKeeperReconciler(
+                RECONCILIATION,
+                vertx,
+                CO_CONFIG,
+                supplier,
+                PFA,
+                patchedKafka,
+                versionChange,
+                null,
+                0,
+                CLUSTER_CA,
+                stateManager.isRollingBack()));
+
+        KafkaStatus status = new KafkaStatus();
+
+        Checkpoint async = context.checkpoint();
+
+        zookeeperReconciler.reconcile(status, Clock.systemUTC()).onComplete(context.succeeding(res -> context.verify(() -> {
+            assertThat(zookeeperReconciler.isKRaftMigrationRollback, is(false));
+            verify(zookeeperReconciler, times(0)).maybeDeleteControllerZnode();
+            async.flag();
+        })));
+    }
+
+    static class MockZooKeeperReconciler extends ZooKeeperReconciler   {
+        public MockZooKeeperReconciler(Reconciliation reconciliation, Vertx vertx, ClusterOperatorConfig config, ResourceOperatorSupplier supplier, PlatformFeaturesAvailability pfa, Kafka kafkaAssembly, KafkaVersionChange versionChange, Storage oldStorage, int currentReplicas, ClusterCa clusterCa, boolean kraftMigrationRollback) {
+            super(reconciliation, vertx, config, supplier, pfa, kafkaAssembly, versionChange, oldStorage, currentReplicas, clusterCa, kraftMigrationRollback);
+        }
+
+        @Override
+        public Future<Void> reconcile(KafkaStatus kafkaStatus, Clock clock)    {
+            return super.isKRaftMigrationRollback ? maybeDeleteControllerZnode() : Future.succeededFuture();
```

Comment:
If you apply my suggestion above, this override will just be ...

```java
maybeDeleteControllerZnode();
```

The current approach is dangerous because if we change the logic in the `ZooKeeperReconciler` we need to change it here as well and we can easily forget. By wrapping everything in the "maybe" method it can't happen.